desk/app/verifier.hoon

```diff
@@ -0,0 +1,1151 @@
+    [%'GET' (make-link turf) ~ ~]
+  --
+::
+++  validate-signature
+  |=  [=bowl:gall sign=(signed)]
+  ^-  ?
+  ::TODO  only count as legit if signature is with latest known life?
+  ::  if we don't know the current life of the signer,
+  ::  or the life used to sign is beyond what we know,
+  ::  we can't validate locally.
+  ::
+  =+  .^(lyf=(unit life) %j /(scot %p our.bowl)/lyfe/(scot %da now.bowl)/(scot %p who.sign))
+  ?~  |(?=(~ lyf) (gth lyf.sign u.lyf))  |
+  ::  jael should have the pubkey. get it and validate.
+  ::
+  =+  .^([life =pass *] %j /(scot %p our.bowl)/deed/(scot %da now.bowl)/(scot %p who.sign)/(scot %ud lyf.sign))
+  (safe:as:(com:nu:crub:crypto pass) sig.sign (jam dat.sign))
+::
+++  revoke
+  |=  $:  [state =bowl:gall]
+          [id=identifier rec=(unit record)]
+          why=@t
+      ==
+  =*  state  +<-<
+  ^-  (quip card _state)
+  =.  rec  ?^(rec rec (~(get by records) id))
+  ?~  rec  [~ state]
+  %-  (~(tell l our.bowl `for.u.rec `-.id) %info ['registration revoked']~)
+  :-  [(give-status for.u.rec id why %gone ~)]~
+  =?  attested  ?=(%done -.status.u.rec)
+    %.  sig.full.status.u.rec
+    %~  del  by
+    (~(del by attested) sig.half.status.u.rec)
+  %_  state
+    records  (~(del by records) id)
+    owners   (~(del ju owners) for.u.rec id)
+    lookups  %+  roll  ~(tap in (~(get ju reverse) id))
+             |=([l=@ =_lookups] (~(del by lookups) l))
+    reverse  (~(del by reverse) id)
+  ==
+::
+++  register
+  |=  $:  [state =bowl:gall]
+          [id=identifier rec=record]
+          proof=(unit proof)
+      ==
+  =*  state  +<-<
+  ^-  (quip card _state)
+  =/  tat=attestation
+    (attest [our now]:bowl for.rec id proof)
+  =.  status.rec  [%done tat]
+  %-  (~(tell l our.bowl `for.rec `-.id) %info ['registration completed']~)
+  :-  [(give-status for.rec id 'registration completed' status.rec)]~
+  %_  state
+    records   (~(put by records) id rec)
+    attested  (~(gas by attested) sig.half.tat^id sig.full.tat^id ~)
+  ==
+::
+++  sign
+  |*  [[our=@p now=@da] dat=*]
+  ^-  (signed _dat)
+  =+  =>  [our=our now=now ..lull]  ~+
+      ;;(=seed:jael (cue .^(@ %j /(scot %p our)/vile/(scot %da now))))
+  ?>  =(who.seed our)
+  =/  sig=@ux  (sigh:as:(nol:nu:crub:crypto key.seed) (jam dat))
+  [our lyf.seed dat sig]
+::
+++  attest
+  |=  [[our=@p now=@da] for=@p id=identifier proof=(unit proof)]
+  ^-  attestation
+  :-  (sign [our now] `half-sign-data`[%0 %verified %half now for -.id])
+  (sign [our now] `full-sign-data`[%0 %verified %full now for id proof])
+::
+++  get-allowance
+  |=  [[solo=(map @p allowance) pool=allowance] for=@p now=@da]
+  |^  ^-  [solo=allowance pool=(unit allowance)]
+      =/  risk=?
+        ?=(?(%pawn %earl) (clan:title for))
+      =/  base=allowance
+        *allowance
+      =/  max  base
+      :-  ?~  lim=(~(get by solo) for)
+            base(since now)
+          (step-allowance u.lim max now (sub now since.u.lim))
+      ?.  risk  ~
```

Comment:
does this mean comet/moon requests don't take from pool?

Reply:
New tests caught bugged refill rate logic!